### ISSUE TITLE
Adding a device while container is running leaves it in a failed state.

### ISSUE THREAD

**Issue by lhorace** [lxc/lxd#5921]
# Required information

 * Distribution: openSUSE
 * Distribution version: tumbleweed
 * The output of "lxc info" or if that fails:
   
```
config:
  core.https_address: '[::]'
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - 10.0.1.47:8443
  - 10.0.3.60:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFYTCCA0mgAwIBAgIQbFGsjT7WdyF+++2ViYbE0DANBgkqhkiG9w0BAQsFADA7
    MRwwGgYDVQQKExNsaW51eGNvbnRhaW5lcnMub3JnMRswGQYDVQQDDBJyb290QGxp
    bnV4LWRlc2t0b3AwHhcNMTkwMzI4MTM0NDMwWhcNMjkwMzI1MTM0NDMwWjA7MRww
    GgYDVQQKExNsaW51eGNvbnRhaW5lcnMub3JnMRswGQYDVQQDDBJyb290QGxpbnV4
    LWRlc2t0b3AwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQDmHFehF1Co
    hf4YT4XW1bbjS1i5Z2ikjCnxUc1A2/K1u++ED7En3oGvc7aqjjMiKdkeKy15FM7T
    MdW8JvogwN2ov2SUgdD5QrBfG8mm54Kgbsa1xRqtG6GqZyDMCEhMy8UezD3Uf5LP
    J7QCf7VcGzSbhWTJ02e8dOGhjxuXq6LIJnckgF22tNVl9DG6I1W5neORGjHbtTKI
    GzJR71pOQAawasuj/YEtgIq4BPMYWcLkhoaWz+9uhxvaJBWEC3SRrHmQbfA6F2+G
    hOkG+RhJ5s2aYRrXUhJTvbn0tojrgUi43gEqSMB3iE2XFRLB+S/v7FP5EyWJU/mf
    I6yQwsD67dH8HPdc7GOdqFt1GEf3lJSosbcc0ZiSVCmHIm/79SjMT9RgpxnYSSkw
    /aRBc+7rLJU5wR5K+UWqmi6Wl667OEXwsaTEfGmtne0OR1pXWhqcp8a+khXcZ9zF
    mz3Uucx5aSF44lf4OMt83ZRRsSdOZ6er0lXAek5nwPc6+gC2lohvQmuu6tIycMLc
    lITUBnXJVvwQdF73tft5JsVi3RBAv+o3BDH1a4ni6yvkihz1rinBFnwVHQZi1pbw
    AsUta1GN9BpYeByfGCwuTi3xdMGu5Y2vekLD/DzNmUHzQarqLnrSFBT4WYU2e/W2
    qsnkvWolrOAfiOib7CT7K2azSpGw0jjOOwIDAQABo2EwXzAOBgNVHQ8BAf8EBAMC
    BaAwEwYDVR0lBAwwCgYIKwYBBQUHAwEwDAYDVR0TAQH/BAIwADAqBgNVHREEIzAh
    gg1saW51eC1kZXNrdG9whwQKAAEvhwQKAAM8hwTAqHoBMA0GCSqGSIb3DQEBCwUA
    A4ICAQBRLSjAZQux+Bg02hbfgHNIirPo2uBjMIw3FsSGDVt7rWFmxdpZyybTgyHk
    E+a3GZt4pT9c8CTexP5Td9biaT3sw99uI4bExcurQNGBUaNY94B31lmQFVxvthrP
    +Ov5J7B8+hnPP+Im9iK0aGZNQT6q2oyl+Hpy0M6tpmVT8p+3kcvtmcUVTw8d/l5y
    b3q2uY6gJQLHgYP9AfVFLF8b2PyeCixsu5PT7vtfVuefgFPb/Q5qZvmRa8T6Min3
    ECSRgj8h64mSs/xvxUMFdkuc3cGiFYnpN/x6nZsPizPiXZgMUi9bdN1tV5/M1V5d
    WQhK7gCpPs0hZ6FdGLblJvXGlBRKPKwt1FlIyln94tyeWvACLjvnly1alBZmZSVV
    VXw1W2MtghkV1G1zbsF8d4fYHxtvpLblipS/nqMuTO0WeB2gweGGVNAZx8j+y8QH
    8c1+NigL7VVGCqGbrmxdwOoaYBUYDtT1JBCz4bLf6yKqW6shz+KawL1z304ao0EU
    8lSJrbqTtnrPI8RXzTqhRs64V5fYHUmY1ghAziMBUxcj26ryun7bznF1JkOJA7rD
    Gun+VRK6rmjm425QX4/mpqb6DM00Cd8RJ/dai1cs3rpLr2sNeN5/aD6k0GlzqrCy
    sdYhDVrsfTXIOWo/1WZgZywaNksm2pVwhWZ3TYfF/AspPkdXgA==
    -----END CERTIFICATE-----
  certificate_fingerprint: 40649f6d73d904fdee6bd2f36c73f13f4a6b895d87b428b6d996e0b58f40f7a2
  driver: lxc
  driver_version: 3.1.0
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    shiftfs: "false"
    uevent_injection: "true"
    unpriv_fscaps: "true"
  kernel_version: 5.1.10-1-default
  lxc_features:
    mount_injection_file: "true"
    network_gateway_device_route: "false"
    network_ipvlan: "false"
    network_l2proxy: "false"
    network_phys_macvlan_mtu: "false"
    seccomp_notify: "false"
  project: default
  server: lxd
  server_clustered: false
  server_name: linux-iqu9
  server_pid: 2252
  server_version: "3.14"
  storage: btrfs
  storage_version: "4.4"
```
# Issue description

I tried to add a tun device to a running unprivilege container and it failed. Shutting down the container was the recourse to successfully add the device. However, it left the container in a broken state in which I had to manually umount /var/snap/lxd/common/lxd/shmounts/ubuntu1904-dev/.lxcmount_sjsu and delete the parent directory to recover.

# Steps to reproduce

 1. Launch a Ubuntu 19.04 container
 2. lxc start ubuntu
 3. lxc config device add ubuntu tun...
 4. lxc shell ubuntu
 5. shutdown -h now
 6. lxc config device add ubuntu tun...
 7. lxc start ubuntu

```
lxc ubuntu1904-dev 20190708231507.614 ERROR    utils - utils.c:safe_mount:1187 - Invalid argument - Failed to mount "/var/snap/lxd/common/lxd/shmounts/ubuntu1904-dev" onto "/var/snap/lxd/common/lxc//dev/.lxd-mounts"
lxc ubuntu1904-dev 20190708231507.614 ERROR    conf - conf.c:mount_entry:2044 - Invalid argument - Failed to mount "/var/snap/lxd/common/lxd/shmounts/ubuntu1904-dev" on "/var/snap/lxd/common/lxc//dev/.lxd-mounts"
lxc ubuntu1904-dev 20190708231507.615 ERROR    conf - conf.c:lxc_setup:3664 - Failed to setup mount entries
lxc ubuntu1904-dev 20190708231507.615 ERROR    start - start.c:do_start:1282 - Failed to setup container "ubuntu1904-dev"
lxc ubuntu1904-dev 20190708231507.615 ERROR    sync - sync.c:__sync_wait:62 - An error occurred in another process (expected sequence number 5)
lxc ubuntu1904-dev 20190708231507.130 ERROR    lxccontainer - lxccontainer.c:wait_on_daemonized_start:864 - Received container state "ABORTING" instead of "RUNNING"
lxc ubuntu1904-dev 20190708231507.130 ERROR    start - start.c:__lxc_start:1975 - Failed to spawn container "ubuntu1904-dev"
lxc ubuntu1904-dev 20190708231507.155 WARN     conf - conf.c:lxc_map_ids:2970 - newuidmap binary is missing
lxc ubuntu1904-dev 20190708231507.155 WARN     conf - conf.c:lxc_map_ids:2976 - newgidmap binary is missing
lxc 20190708231507.158 WARN     commands - commands.c:lxc_cmd_rsp_recv:132 - Connection reset by peer - Failed to receive response for command "get_state"
```

**Comment by stgraber** [comment]
Don't use LXD on a 5.1 kernel which didn't get @brauner' fix to mount handling or it will fail badly and may even end up with data loss.

**Comment by stgraber** [comment]
See https://github.com/lxc/lxc/issues/3082 for some more details on this
The issue was fixed upstream and should hit stable for the 5.1 kernel too.

**Comment by lhorace** [comment]
I didn't get a chance to catch up with E-Mail but I'm bit confused and I will comment on #3082
